aws_lambda_powertools/utilities/data_classes/kinesis_firehose_event.py

```diff
@@ -0,0 +1,113 @@
+import base64
+import json
+from typing import Iterator, Optional
+
+from aws_lambda_powertools.utilities.data_classes.common import DictWrapper
+
+
+class KinesisFirehoseRecordMetadata(DictWrapper):
+    @property
+    def _metadata(self) -> dict:
+        """Optional: metadata associated with this record; present only when Kinesis Stream is source"""
+        return self["kinesisRecordMetadata"]  # could raise KeyError
+
+    @property
+    def shard_id(self) -> Optional[str]:
+        """Kinesis stream shard ID; present only when Kinesis Stream is source"""
+        return self._metadata.get("shardId")
+
+    @property
+    def partition_key(self) -> Optional[str]:
+        """Kinesis stream partition key; present only when Kinesis Stream is source"""
+        return self._metadata.get("partitionKey")
+
+    @property
+    def approximate_arrival_timestamp(self) -> Optional[str]:
+        """Kinesis stream approximate arrival ISO timestamp; present only when Kinesis Stream is source"""
+        return self._metadata.get("approximateArrivalTimestamp")
+
+    @property
+    def sequence_number(self) -> Optional[str]:
+        """Kinesis stream sequence number; present only when Kinesis Stream is source"""
+        return self._metadata.get("sequenceNumber")
+
+    @property
+    def subsequence_number(self) -> Optional[str]:
+        """Kinesis stream sub-sequence number; present only when Kinesis Stream is source
+
+        Note: this will only be present for Kinesis streams using record aggregation
+        """
+        return self._metadata.get("subsequenceNumber")
+
+
+class KinesisFirehoseRecord(DictWrapper):
+    @property
+    def approximate_arrival_timestamp(self) -> float:
+        """The approximate time that the record was inserted into the delivery stream"""
+        return float(self["approximateArrivalTimestamp"])
+
+    @property
+    def record_id(self) -> str:
+        """Record ID; uniquely identifies this record within the current batch"""
+        return self["recordId"]
+
+    @property
+    def data(self) -> str:
+        """The data blob, base64-encoded"""
+        return self["data"]
+
+    @property
+    def metadata(self) -> KinesisFirehoseRecordMetadata:
+        """Optional: metadata associated with this record; present only when Kinesis Stream is source"""
+        return KinesisFirehoseRecordMetadata(self._data)
```

Comment:
It looks like there's no metadata when the Kinesis Firehose is configured with "Direct PUT" mode. As you mention in the comment, this field is optional. So can we mark this field as optional too?

```suggestion
    def metadata(self) -> Optional[KinesisFirehoseRecordMetadata]:
        """Optional: metadata associated with this record; present only when Kinesis Stream is source"""
        return KinesisFirehoseRecordMetadata(self._data) if self.get("kinesisRecordMetadata") else None
```

Reply:
+1 for mark this as optional

Reference: https://github.com/awslabs/aws-lambda-powertools-python/pull/1556#pullrequestreview-1132125971

Reply:
done